storage/reads/array_cursor.go

```diff
@@ -18,84 +17,52 @@ func (v *singleValue) Value(key string) (interface{}, bool) {
+func newWindowAggregateArrayCursor(ctx context.Context, agg *datatypes.Aggregate, every int64, cursor cursors.Cursor) cursors.Cursor {
 	if cursor == nil {
 		return nil
 	}
 
-	switch req.Aggregate[0].Type {
+	switch agg.Type {
 	case datatypes.AggregateTypeCount:
-		return newWindowCountArrayCursor(cursor, req)
+		return newWindowCountArrayCursor(cursor, every)
+	case datatypes.AggregateTypeSum:
+		return newWindowSumArrayCursor(cursor, every)
 	default:
 		// TODO(sgc): should be validated higher up
 		panic("invalid aggregate")
 	}
 }
 
-func newSumArrayCursor(cur cursors.Cursor) cursors.Cursor {
+func newWindowCountArrayCursor(cur cursors.Cursor, every int64) cursors.Cursor {
 	switch cur := cur.(type) {
 	case cursors.FloatArrayCursor:
-		return newFloatArraySumCursor(cur)
+		return newFloatWindowCountArrayCursor(cur, every)
 	case cursors.IntegerArrayCursor:
-		return newIntegerArraySumCursor(cur)
+		return newIntegerWindowCountArrayCursor(cur, every)
 	case cursors.UnsignedArrayCursor:
-		return newUnsignedArraySumCursor(cur)
-	default:
-		// TODO(sgc): propagate an error instead?
-		return nil
-	}
-}
-
-func newCountArrayCursor(cur cursors.Cursor) cursors.Cursor {
-	switch cur := cur.(type) {
-	case cursors.FloatArrayCursor:
-		return newFloatCountArrayCursor(cur)
-	case cursors.IntegerArrayCursor:
-		return newIntegerCountArrayCursor(cur)
-	case cursors.UnsignedArrayCursor:
-		return newUnsignedCountArrayCursor(cur)
+		return newUnsignedWindowCountArrayCursor(cur, every)
 	case cursors.StringArrayCursor:
-		return newStringCountArrayCursor(cur)
+		return newStringWindowCountArrayCursor(cur, every)
 	case cursors.BooleanArrayCursor:
-		return newBooleanCountArrayCursor(cur)
+		return newBooleanWindowCountArrayCursor(cur, every)
 	default:
 		panic(fmt.Sprintf("unreachable: %T", cur))
 	}
 }
 
-func newWindowCountArrayCursor(cur cursors.Cursor, req *datatypes.ReadWindowAggregateRequest) cursors.Cursor {
-	if req.WindowEvery == math.MaxInt64 {
-		// This means to aggregate over the entire range,
-		// don't do windowed aggregation.
-		return newCountArrayCursor(cur)
-	}
+func newWindowSumArrayCursor(cur cursors.Cursor, every int64) cursors.Cursor {
 	switch cur := cur.(type) {
 	case cursors.FloatArrayCursor:
-		return newFloatWindowCountArrayCursor(cur, req.WindowEvery)
+		return newFloatWindowSumArrayCursor(cur, every)
 	case cursors.IntegerArrayCursor:
-		return newIntegerWindowCountArrayCursor(cur, req.WindowEvery)
+		return newIntegerWindowSumArrayCursor(cur, every)
 	case cursors.UnsignedArrayCursor:
-		return newUnsignedWindowCountArrayCursor(cur, req.WindowEvery)
-	case cursors.StringArrayCursor:
-		return newStringWindowCountArrayCursor(cur, req.WindowEvery)
-	case cursors.BooleanArrayCursor:
-		return newBooleanWindowCountArrayCursor(cur, req.WindowEvery)
+		return newUnsignedWindowSumArrayCursor(cur, every)
 	default:
-		panic(fmt.Sprintf("unreachable: %T", cur))
+		panic(fmt.Sprintf("unsupported for aggregate sum: %T", cur))
```

Comment:
Is this strictly unreachable? As this effort matures, it might make sense to return error with the cursor here and above. Just a nit today.

Reply:
Yeah, I think we need better error handling all around here. It's tricky to return an error here because the `Cursor()` method is part of an interface, so the interface and all of its implementors would need to be updated. I'd like to defer that work to a later time.